Store TrackExtra, TrackingRecHits, and SiPixel/StripClusters for muons in AOD+MINIAOD (Self-contained version without framework thinning)

bendavid opened the pull request:
Precision measurements using muons might want to be able to refit the muon tracks (and the momentum scale corrections for W mass are moving in this direction).

This PR adds thinned collections of TrackExtra and the corresponding TrackingRecHits and Strip/Pixel clusters to AOD for those tracks which are associated to muon::bestTrack() with some cuts applied to the muons.
(this means that for most muons with pt<200GeV only the pixel and strip rechits are stored and not the muon chamber ones, because bestTrack points to the inner track unless the muon is purely standalone). This behaviour could be modified if desired to store some additional trackExtra/hits for the global track.

The cut applied to the muons to store the extra information in AOD is pt>3.0 GeV.

A further reduced set of collections is stored in MINIAOD, applying a 4.5 GeV cut on top of the selectedPatMuons (such that the extra cuts in https://github.com/cms-sw/cmssw/blob/ee9e9ae286520cc8c458ac991da3f4b0c7fbea75/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L118 are implicitly included)

This is a fallback alternative to https://github.com/cms-sw/cmssw/pull/30544 in case the needed framework functionality is not integrated quickly enough to backport this to 10_6 in time for the restart of the UL MC campaigns.  (Since the AOD event content is modified it would be highly desirable to have this in for consistency purposes)

This PR does not touch any existing producers in the RECO sequence and only adds additional collections.  Small modifications are made to the PATMuonSlimmer in the MINIAOD sequence.

Tested using 11834.0 (TTbar 2021 realistic with PU Run3_Flat55To75_PoissonOOTPU) 500 events.

The additional collections in AOD are:
```
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
recoTrackExtras_muonReducedTrackExtras__RECO. 951.718 619.05
SiStripClusteredmNewDetSetVector_muonReducedTrackExtras__RECO. 704.88 312.796
SiPixelClusteredmNewDetSetVector_muonReducedTrackExtras__RECO. 521.332 288.952
TrackingRecHitsOwned_muonReducedTrackExtras__RECO. 2374.61 268.714
recoTrackExtrasedmAssociation_muonReducedTrackExtras__RECO. 9695.24 135.65
```
(and the total size per event increases from 672 kB to 674 kB for a 0.25% increase)


The additional collections in MINIAOD are:
```
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
TrackingRecHitsOwned_slimmedMuonTrackExtras__RECO. 1919.71 387.802
recoTrackExtras_slimmedMuonTrackExtras__RECO. 639.646 303.8
SiStripClusteredmNewDetSetVector_slimmedMuonTrackExtras__RECO. 368.938 157.746
SiPixelClusteredmNewDetSetVector_slimmedMuonTrackExtras__RECO. 235.028 100.002
```
(and the total size per event increases from 101.5 kB to 102.4 kB for a 0.9% increase)

Some drawbacks of this approach with respect to the leveraging of framework thinning and slimming as in https://github.com/cms-sw/cmssw/pull/30544:
-code is based on recursive rekeying and is therefore harder to maintain
-some duplication of reco::TrackExtras which are already stored in AOD (but need to be duplicated in order to have consistent rekeying of references to TrackingRecHits in the reduced collection)
-additional information is not easily accessible at AOD level, because the final reference rekeying only happens during the production of the slimmedMuons at MINIAOD level (but the association maps are present, so it is possible at least)

Some open points with this PR:
1) information stored is incomplete for phase 2 (but phase 2 workflows should still run)
2) edm::Association is maybe not the most efficient for this purpose (see large uncompressed size, though it compresses fairly well given that most of the references for generalTracks are null)
3) ```edm::Association<reco::TrackExtraCollection>``` required adding additional dictionaries in DataFormats/TrackReco (so this will produce errors when reading the corresponding AOD files with an older release unless dropping those on input)
4) Currently the use of the additional information at the MINIAOD step is DISABLED for Run 2 UL, since the needed collections are not present in the AOD.  If we manage to get this in before the restart of production in the latest 10_6 release, then the process modifiers for Run 2 UL will need to be split into "new" and "old" variants, with this disabled for Run2ULold, but left enabled for Run2ULnew

Point 1) I think can be dealt with in a future PR
For points 2) and 3) it may be ok as is, but open to suggestions for alternatives to ```edm::Association<reco::TrackExtraCollection>``` for this use case (associating the original TrackExtra references to the reduced collection)

bendavid committed:
add dictionaries for edm::Association<reco::TrackExtraCollection>

bendavid committed:
add MuonReducedTrackExtraProducer

bendavid committed:
Add reduced TrackExtras, TrackingRecHits, and SiPixel/StripClusters for muon tracks to RECO sequence and AOD event content

bendavid committed:
add function to rekey TrackExtra references of embedded tracks

bendavid committed:
add optional rekeying of TrackExtra references of embedded tracks

bendavid committed:
Add reduced TrackExtras, TrackingRecHits, and SiPixel/StripClusters for muon tracks to MINIAOD sequence and event content

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/17870

- This PR adds an extra 68KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30898, #31129
  - File  PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py  modified in PR(s): #30898
  - File  PhysicsTools/PatAlgos/python/slimming/slimmedTrackExtras_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #31129
  - File  RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py  modified in PR(s): #30544
  - File  RecoMuon/Configuration/python/RecoMuon_cff.py  modified in PR(s): #30544
  - File  RecoMuon/MuonIdentification/plugins/Module.cc  modified in PR(s): #30544


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/17870/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/17870/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

bendavid committed:
code style fixes for MuonReducedTrackExtraProducer

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/17871

- This PR adds an extra 68KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30898, #31129
  - File  PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py  modified in PR(s): #30898
  - File  PhysicsTools/PatAlgos/python/slimming/slimmedTrackExtras_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #31129
  - File  RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py  modified in PR(s): #30544
  - File  RecoMuon/Configuration/python/RecoMuon_cff.py  modified in PR(s): #30544
  - File  RecoMuon/MuonIdentification/plugins/Module.cc  modified in PR(s): #30544

cmsbuild commented:
A new Pull Request was created by  @bendavid (Josh Bendavid) for master.

It involves the following packages:

DataFormats/PatCandidates
DataFormats/TrackReco
PhysicsTools/PatAlgos
RecoMuon/Configuration
RecoMuon/MuonIdentification


@perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @gouskos, @abbiendi, @emilbols, @echapon, @ahinzmann, @peruzzim, @seemasharmafnal, @mmarionncern, @JyothsnaKomaragiri, @makortel, @smoortga, @jhgoh, @jdolen, @HuguesBrun, @cericeci, @ferencek, @trocino, @rociovilar, @rovere, @VinInn, @jdamgov, @bellan, @nhanvtran, @gkasieczka, @schoef, @andrzejnovak, @clelange, @riga, @Fedespring, @calderona, @hatakeyamak, @cbernet, @gpetruc, @mariadalfonso, @folguera this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

VinInn commented:
I may have missing some details. Would not be easier and more robust to just make a deep copy?

bendavid commented:
Deep copy of what?

The needed information is spread over reco::Track, reco::TrackExtra, TrackingRecHit, SiPixelCluster, SiStripCluster, and these objects are linked together by edm::Ref, so the only two possibilities I'm aware of are
1) Separately copy each collection and recursively rekey (as done here)
2) Use framework thinning functionality to separately copy each collection and rely on the framework to forward the references automatically (as done in https://github.com/cms-sw/cmssw/pull/30544 but needing also some non-trivial extensions to the framework thinning functionality to be viable)

Open to other alternatives of course.

VinInn commented:
track, extra and keys can be deep copied using 
https://cmssdt.cern.ch/lxr/source/RecoTracker/FinalTrackSelectors/src/TrackCollectionCloner.cc?v=CMSSW_11_2_X_2020-08-19-2300

so what is left is to copy clusters and to rekey omni-refs in hits
(I assume one does not care about seeds)

I admit that DetSetVector is a pretty complex container for thinning and a simpler vector<SiPIxelCluster> and vector<SiStripCluster> would be easier to fill and handle...
BUT the Ref to cluster is strongly related to DetSetVector even if at the end it is just locating the cluster in a vector<XYClusters>

makortel commented:
> I admit that DetSetVector is a pretty complex container for thinning and a simpler vector and vector would be easier to fill and handle...

`edmNew::DetSetVector` wasn't really that complicated in the end (at least for the "thinning" in framework), and https://github.com/cms-sw/cmssw/pull/31100 supports it.

bendavid commented:
Yes this is more or less what MuonReducedTrackExtraProducer in this PR is doing (in this case we actually don't want to copy the tracks, because they are all already stored in AOD in the full collections, and they get copied at the MINIAOD step anyways when they are embedded in the pat::Muons)

VinInn commented:
if a vector<XYCluster> would have been sufficient one could have avoided all the code in
https://github.com/cms-sw/cmssw/pull/31217/files#diff-25b6ffc7f1a7fb0ddc2a61dd171b1a06R177
and just push the cluster in its vector in the previous loop and make a trivial "rekey" 
(btw no matched hits shall be present in standard offline wf at that stage)

bendavid commented:
Yes I wanted to anyways handle the case that a cluster is referenced multiple times.  (It's possible this code could be extended in the future to store multiple variations of the same track)

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 8f14f6b761555105a4f8031980c2cf42524ff299

CMSSW: CMSSW_11_2_X_2020-08-24-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/8894/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests  RelVals  AddOn**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test runtestTqafTopEventSelection had ERRORS
---> test runtestTqafExamples had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
---> test runtestTqafTopHitFit had ERRORS
---> test runtestTqafTopJetCombination had ERRORS
---> test runtestTqafTopKinFitter had ERRORS
---> test runtestTqafTopTools had ERRORS


* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>1000.0 step3
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step3_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>23234.0 step3
<pre>runTheMatrix-results/23234.0_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/step3_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log

</pre>28234.0 step3
<pre>runTheMatrix-results/28234.0_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal/step3_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal.log

</pre>23434.999 step4
<pre>runTheMatrix-results/23434.999_TTbar_14TeV+2026D49PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU/step4_TTbar_14TeV+2026D49PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py TTbar_8TeV_TuneCUETP8M1_cfi  --conditions auto:run1_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot Realistic8TeVCollision : FAILED - time: date Tue Aug 25 00:28:14 2020-date Tue Aug 25 00:23:34 2020 s - exit: 17920
cmsRun /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-08-24-1100/src/PhysicsTools/PatAlgos/test/IntegrationTest_cfg.py : FAILED - time: date Tue Aug 25 00:24:04 2020-date Tue Aug 25 00:23:38 2020 s - exit: 18688
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc_l1stage1 --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_25ns : FAILED - time: date Tue Aug 25 00:29:30 2020-date Tue Aug 25 00:23:50 2020 s - exit: 17920
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 : FAILED - time: date Tue Aug 25 00:29:39 2020-date Tue Aug 25 00:23:55 2020 s - exit: 17920

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

bendavid committed:
add missing parameter

bendavid committed:
skip clusters for fastsim and phase2 cases

bendavid commented:
Failing tests should be fixed now.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/17885

- This PR adds an extra 72KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30898, #31129
  - File  PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py  modified in PR(s): #30898
  - File  PhysicsTools/PatAlgos/python/slimming/slimmedTrackExtras_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #31129
  - File  RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py  modified in PR(s): #30544
  - File  RecoMuon/Configuration/python/RecoMuon_cff.py  modified in PR(s): #30544
  - File  RecoMuon/MuonIdentification/plugins/Module.cc  modified in PR(s): #30544

cmsbuild commented:
Pull request #31217 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 65676248e7818dddd574e055dd4aa220b762b0aa
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/8901/summary.html
CMSSW: CMSSW_11_2_X_2020-08-24-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/8901/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2609656
 - DQMHistoTests: Total failures: 11
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2609622
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

bendavid commented:
For what concerns the ```edm::Association<reco::TrackExtraCollection>``` I don't see an alternative with dictionaries already existing, and other alternatives would increase the complexity of the code (e.g. edm::AssociationMap which would require separating the various track extra collections) without necessarily being any more efficient, so I would propose to leave this as is.

For what concerns the process modifiers, I would also propose to leave things as they are for this PR (disabled for run 2 UL) and this can be updated once the new UL campaigns with this PR backported and updated AOD content actually exist.

perrotta commented:
> This is a fallback alternative to #30544 in case the needed framework functionality is not integrated quickly enough to backport
> this to 10_6 in time for the restart of the UL MC campaigns. (Since the AOD event content is modified it would be highly desirable  to have this in for consistency purposes)

@bendavid , could you please specify better the purpose of this PR?

If #30544 + #31100 will become ready "in time" this can be simply closed and removed? Then, when such "in time" will be?

Since you want to backport this feature, assuming #30544 + #31100 will get eventually merged in the master, do you plan to backport the two of them, or still the version coded in this PR? In view of such a possible backport, do you plan to integrate this one first, then backport it, and then #30544 + #31100, opportunely rebased, in the master?

bendavid commented:
I believe that ideally PPD would like to open the new UL campaigns ("Summer20") on the timescale of later next week, with a new 10_6 release including this feature in some form.

While there are reasons why #30544 + #31100 would be preferred (listed in the description of this PR), I'm afraid that there are/will be still open points and/or that the backport may be non-trivial.  So it's probably more realistic to merge this PR, then backport it to 10_6.   Then #30544 can be reworked to adapt the functionality to use the thinning in 11X (but there will be no need to backport #31100 to 10_6)

rappoccio commented:
From our standpoint, earlier is better here. We're ready to launch Summer20 (or, we were, but are waiting for this PR). The Wmass measurement is a very high priority for us so we would like this in the queue ASAP.

perrotta commented:
Thank you @bendavid and @rappoccio 
Ok, now there is a plan, then. I will concentrate on the review of this PR during the next days.

perrotta commented on RecoMuon/MuonIdentification/plugins/MuonReducedTrackExtraProducer.h:
```diff
@@ -0,0 +1,49 @@
+  class Muon;
+}
+
+class MuonReducedTrackExtraProducer : public edm::stream::EDProducer<> {
```
Couldn't this become a `global` module?

bendavid replied:
Are you sure StringCutObjectSelector is thread-safe?

perrotta replied:
Uhm, there are examples of global producers in which there is a  `StringCutObjectSelector` as a class member. But honestly I'm not sure it is thread safe, and browsing around I also found authoritative statements that claim it is not.
In any case, I was just guessing: no reason to spend time in investigating it if it doesn't turn out immediate

perrotta replied:
Thinking at it a bit more, I don't see why StringCutObjectSelector as used here cannot be thread safe.
@Dr15Jones , what do you think?

Dr15Jones replied:
I'm not confident StringCutObjectSelector is thread-safe, especially since it calls down into ROOT. We know it to be thread-friendly.

perrotta replied:
Ok, thank you @Dr15Jones 
By the way, do you see a possible issue in (e.g.) ​PhysicsTools/​PatAlgos/​plugins/​PATLostTracks.cc?

Dr15Jones replied:
In `PATLostTracks` the `StringCutObjectSelector` is make in the constructor and is set to be _non lazy_ which avoids calling ROOT's cling during event processing. Unfortunately one part of the Selector system isn't thread safe: ExpressionVar. It has two `mutable` method it modifies during expression evaluation:

https://github.com/cms-sw/cmssw/blob/master/CommonTools/Utils/src/ExpressionVar.h#L26-L27

perrotta commented on RecoMuon/MuonIdentification/plugins/MuonReducedTrackExtraProducer.cc:
```diff
@@ -0,0 +1,253 @@
+      trackExtraOutToken_(produces<reco::TrackExtraCollection>()),
+      trackingRecHitsOutToken_(produces<TrackingRecHitCollection>()),
+      associationOutToken_(produces<edm::Association<reco::TrackExtraCollection>>()) {
+  std::vector<edm::InputTag> trackExtraTags = pset.getParameter<std::vector<edm::InputTag>>("trackExtraTags");
```
`const`

bendavid committed:
move rekeying of TrackExtra references to PATMuonSlimmer

bendavid commented on RecoMuon/MuonIdentification/plugins/MuonReducedTrackExtraProducer.h:
```diff
@@ -0,0 +1,49 @@
+  edm::EDPutTokenT<TrackingRecHitCollection> trackingRecHitsOutToken_;
+  edm::EDPutTokenT<edmNew::DetSetVector<SiPixelCluster>> pixelClusterOutToken_;
+  edm::EDPutTokenT<edmNew::DetSetVector<SiStripCluster>> stripClusterOutToken_;
+  edm::EDPutTokenT<edm::Association<reco::TrackExtraCollection>> associationOutToken_;
```
Sorry what is the change you're suggesting here?

perrotta replied:
Make it `const` (sorry, I pasted some wrong string in my previous comment)

bendavid committed:
make data members const where possible

bendavid committed:
adapt MINIAOD sequence and event content to accommodate two levels of reduction for muon track extras

bendavid committed:
remove default destructor

bendavid committed:
simplify application of muon cuts

bendavid committed:
extend rekeying of TrackExtraRefs to allow rekeying through multiple levels of reduction

bendavid committed:
Extend MuonReducedTrackExtraProducer to allow multiple levels of reduction

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/18014

- This PR adds an extra 120KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30898, #31129
  - File  PhysicsTools/PatAlgos/python/slimming/slimmedTrackExtras_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #31129
  - File  RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py  modified in PR(s): #30544
  - File  RecoMuon/Configuration/python/RecoMuon_cff.py  modified in PR(s): #30544
  - File  RecoMuon/MuonIdentification/plugins/Module.cc  modified in PR(s): #30544

cmsbuild commented:
Pull request #31217 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

bendavid commented:
Aside from addressing comments, I managed to get things working with two level of reduction without too much added complication, such that the cuts can be different at AOD and MINIAOD level (following some questions from @arizzi).  The PR description has been updated to reflect some changes in the details of the cuts and extra collections.

bendavid commented:
For comparison, if the cuts are lowered to 3GeV, the AOD increases to 0.25% over baseline (vs 0.16% with the 4.5 GeV cut)

The MINIAOD increase goes from 0.9% to 1.4%

(With the latest updates the cuts can be adjusted separately for AOD and MINIAOD)

arizzi commented:
we should check with BPH then if having the muon hits in the MiniAOD could help avoiding the need of AOD for some analyses.

perrotta commented:
> Aside from addressing comments, I managed to get things working with two levels of reduction without too much added complication, such that the cuts can be different at AOD and MINIAOD level (following some questions from @arizzi). The PR description has been updated to reflect some changes in the details of the cuts and extra collections.

Are the size increases reported in the PR description updated with the new thresholds? If I am not wrong (you reshuffled the last commits, and it is therefore difficult to evaluate separately the last changes) there is now a looser selection at AOD, while it remains the same as before your last updates for miniAOD.

Can you also evaluate, and write in the PR description, the cost in terms of CPU? I don't expect it will be relevant, but it would be nice to have it recorded for reference

bendavid commented:
Yes the PR description was updated to reflect the current state of the PR.  The size didn't change much for AOD, because the cuts in effect are only looser between 4.5 and 5GeV.

For MINIAOD indeed the cuts are identical, and the size is very similar, but the collections are slightly different (now the extra content for standAloneMuons is mixed into the slimmedMuonTrackExtras instead of being a separate set of collections)

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: ff0939bef16f98da6299275630678be0dcde5b36

CMSSW: CMSSW_11_2_X_2020-08-30-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/9013/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS

cmsbuild commented:
Comparison job queued.

bendavid commented:
Unit test failures happen because the Tau embedding uses slimmedMuons added to RAWRECO output.

Two possible fixes would be
1) Add the extra slimmed track collections to their custom event content at https://github.com/cms-sw/cmssw/blob/ee9e9ae286520cc8c458ac991da3f4b0c7fbea75/TauAnalysis/MCEmbeddingTools/python/customisers.py#L83-L96

2) Disable rekeying of the TrackExtra refs in the PATMuonSlimmer specifically for the tau embedding workflows, since the full original collections are anyways present in the RAW-RECO output

Need to check whether the tau embedding workflow relies on the TrajectoryStateParameters in the TrackExtra (in which case 1) will not fully solve the issue)

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/9013/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2609667
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2609644
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

bendavid committed:
skip rekeying of TrackExtra refs in tau embedding workflow

bendavid commented:
The failing unit test should be fixed now.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/18035

- This PR adds an extra 132KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30898, #31129
  - File  PhysicsTools/PatAlgos/python/slimming/slimmedTrackExtras_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #31129
  - File  RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py  modified in PR(s): #30544
  - File  RecoMuon/Configuration/python/RecoMuon_cff.py  modified in PR(s): #30544
  - File  RecoMuon/MuonIdentification/plugins/Module.cc  modified in PR(s): #30544

cmsbuild commented:
Pull request #31217 was updated. @perrotta, @civanch, @mdhildreth, @cmsbuild, @slava77, @jpata, @santocch can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: b3249935763f788e55080ac7456f683b62b72d96
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/9020/summary.html
CMSSW: CMSSW_11_2_X_2020-08-31-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/9020/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2609667
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2609644
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

bendavid commented:
For the process modifiers, thinking a bit more, I think the best would be to add a new process modifier

run2_miniAOD_UL_Summer19 which chains on top of the existing run2_miniAOD_UL

Then the changes related to the extra bits here can be attached to that one.

Shall I add that already to this PR?

slava77 commented on PhysicsTools/PatAlgos/python/slimming/slimmedMuons_cfi.py:
```diff
@@ -15,6 +15,19 @@
 )
 
+# full set of track extras not available in existing AOD
+from Configuration.Eras.Modifier_run2_miniAOD_80XLegacy_cff import run2_miniAOD_80XLegacy
+run2_miniAOD_80XLegacy.toModify(slimmedMuons, trackExtraAssocs = ["slimmedMuonTrackExtras"])
+
+from Configuration.Eras.Modifier_run2_miniAOD_94XFall17_cff import run2_miniAOD_94XFall17
+run2_miniAOD_94XFall17.toModify(slimmedMuons, trackExtraAssocs = ["slimmedMuonTrackExtras"])
+
+from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
+pp_on_AA_2018.toModify(slimmedMuons, trackExtraAssocs = ["slimmedMuonTrackExtras"])
+
+from Configuration.ProcessModifiers.run2_miniAOD_UL_cff import run2_miniAOD_UL
+run2_miniAOD_UL.toModify(slimmedMuons, trackExtraAssocs = ["slimmedMuonTrackExtras"])
```
`(run2_miniAOD_80XLegacy | run2_miniAOD_94XFall17 | pp_on_AA_2018 | run2_miniAOD_UL).toModify(...`
will be shorter and more maintainable

it seems like a repeated pattern

perrotta replied:
This comment wasn't taken into account yet.
It is just a simplification, but I'd suggest to consider it already in this PR, for an easier maintenance

bendavid replied:
This was addressed by https://github.com/cms-sw/cmssw/pull/31217/commits/17ae4b93297f1340df88ebf138ad7ca877541fb1 no?

perrotta replied:
> This was addressed by [17ae4b9](https://github.com/cms-sw/cmssw/commit/17ae4b93297f1340df88ebf138ad7ca877541fb1) no?

Ah, ok! Why I still did not see it? Maybe I had open an outdated version on my github...
Now I can see the modification applied, sorry for the noise

slava77 commented:
> For the process modifiers, thinking a bit more, I think the best would be to add a new process modifier
> 
> run2_miniAOD_UL_Summer19 which chains on top of the existing run2_miniAOD_UL
> 
> Then the changes related to the extra bits here can be attached to that one.
> 
> Shall I add that already to this PR?

what's the full context of using this new modifier?
This PR also modifies AOD.
`run2_miniAOD*` should not be used for making AOD.

If it's just for the miniAOD part, I'm not sure what you intent to add in `run2_miniAOD_UL_Summer19`; naively, it sounds like removing the trackExtra inputs.
If that's a super-targeted modifier, I'd suggest that it has a more functional name.

bendavid commented:
Yes, the modifier only applies to the miniaod part, and yes indeed it's super targeted as you say, but since it applies broadly to all of the AOD from the UL data rereco and Summer19 MC campaigns, it would seem good to give it a general name, since it will need to be used for any future reminiaod campaign starting fro those datasets.

slava77 commented:
I'm not sure that a longer abstract name is useful, especially if the thing does only one thing (`skip_miniAOD_trackExtras`).
I'd at least put `_extra` in the name to indicate that it's not a full modifier.

bendavid commented:
Ok but the idea is that this would be a ModifierChain on top of run2_miniAOD_UL such that this would be the overall process modifier to be used when running re-miniaod of the existing UL aod.

This additional piece being the only difference between running remini-aod on older UL AOD vs the new AOD produced from the upcoming Summer20 campaign.  If this is the case then it would be confusing to give it a very specific name.

If it has a specific name then there should be some other modifier build on top of it which combined it with the general run2_miniAOD_UL I suppose.

slava77 commented:
> Ok but the idea is that this would be a ModifierChain

my earlier comments were for individually listed elements in the `--processModifier` arguments.
IIRC, this part does not support ModifierChain.

bendavid commented:
Ok, so your proposal is to set things up such that any reminiaod of Summer19 MC or UL data would be done with
```--procModifiers run2_miniAOD_UL,skip_miniAOD_trackExtras```

That's of course also fine with me.

I DO see an example of ModifierChain in Configuration/ProcessModifiers though:
https://github.com/cms-sw/cmssw/blob/bb3e3988bd41b929f5760e5b052594b56c8672e6/Configuration/ProcessModifiers/python/Input_92xAOD_cff.py#L4

slava77 commented:
> I DO see an example of ModifierChain in Configuration/ProcessModifiers though:
> Input_92xAOD

I do not see a single use case; does it work?
if so, we can go back to the chain with the Summer19 in the name

bendavid committed:
introduce specific process modifier to skip additional trackExtras collections during miniaod, and corresponding ModifierChain for run2_miniAOD_UL_preSummer20

bendavid committed:
simplify use of ProcessModifiers and move to dedicated miniAOD_skip_trackExtras rather than disabling additional trackExtras for run 2 UL in general

cmsbuild commented:
The code-checks are being triggered in jenkins.

bendavid commented:
Yes the ModifierChain seems to work, I put "preSummer20" rather than "Summer19" since it also applies to the data.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/18069

- This PR adds an extra 180KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30898, #31129
  - File  PhysicsTools/PatAlgos/python/slimming/slimmedTrackExtras_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #31129
  - File  RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py  modified in PR(s): #30544
  - File  RecoMuon/Configuration/python/RecoMuon_cff.py  modified in PR(s): #30544
  - File  RecoMuon/MuonIdentification/plugins/Module.cc  modified in PR(s): #30544

cmsbuild commented:
Pull request #31217 was updated. @perrotta, @civanch, @jordan-martins, @chayanit, @wajidalikhan, @mdhildreth, @cmsbuild, @silviodonato, @franzoni, @kpedro88, @slava77, @jpata, @qliphy, @santocch, @fabiocos, @davidlange6 can you please check and sign again.

slava77 reviewed (commented):
IMHO, this last update may become pretty confusing.

It sounds like `run2_miniAOD_UL` for data will work only for the special SMP requests, while for MC it is proposed to be "the final default".

Also, without a full validation campaign, I don't think that we can deploy run2_miniAOD_UL_preSummer20 or run2_miniAOD_UL in 106X (the features are also still incomplete, ETA sounds like end of Sept or so), while it sounds like the goal of #31217 is to introduce this limited thinning by default ASAP.

bendavid commented:
Hi Slava,
Yes, with these changes ```run2_miniAOD_UL``` is the default, and the idea is that it will work for all Summer20 MC samples, and for any special data rereco for W mass.

Indeed any reminiaod of the data or Summer19 MC will have to use the ```run2_miniAOD_UL_preSummer20``` modifier instead.

The idea of this PR is that it doesn't touch any existing producer in the reco sequence, and that it has as minimal as possible changes to the MINIAOD sequence.

The ONLY change to the reco sequence and event content is to run one additional producer and save the extra collections.

I'm not sure what you say is still incomplete here.  This PR does not depend on the additional thinning/slimming functionality on the framework side.

There is a strong motivation to have this included for the Summer20 campaign in general, because it is a natural breakpoint for the event content.  Otherwise we create even more mess/confusion by having special event content only for the W mass samples.

slava77 commented:
> I'm not sure what you say is still incomplete here. This PR does not depend on the additional thinning/slimming functionality on the framework side.
> 
> There is a strong motivation to have this included for the Summer20 campaign in general, because it is a natural breakpoint for the event content. Otherwise we create even more mess/confusion by having special event content only for the W mass samples.

the master version is probably OK, with the caveat of the possible confusion of where to apply UL and where to apply UL_preSummer20.

The point is more for the backport to 106X: IIUC, UL will not be in a production campaign until what looks like October. So, if you want something to start sooner, it would have to be something that adds the extras/hits/clusters without adding the UL modifier.

bendavid commented:
This is all enabled by default, so if no modifier is used then it will be active.

So indeed the idea is that this gets backported to 10_6 as it is now, and the extra functionality stays enabled for the Summer20 campaign.

The MINIAOD step for the Summer19 campaign is simply not going to update to the new release until the reminiaod is ready for production as I understand.  (But if someone really wanted to do that, they could still do it without pulling in the other modifier by using ```--processModifier miniAOD_skip_trackExtras```)

If you have other suggestions for the naming I'm of course open to it.

slava77 commented:
> This is all enabled by default, so if no modifier is used then it will be active.

ah, all ok then (sorry for the noise).

perrotta commented:
I've checked the performance with two different wfs: 10224 (TTbar PU with 2017 settings) and 11834 (TTbar 2021 realistic with PU Run3_Flat55To75_PoissonOOTPU, the same reported in the PR description). Only 100 events each, therefore some different compression factor wrt what written in the PR description for 11834 is expected.

In any case, the results are fully comparable in the 2017 and 2021 settings, both in terms of output sizes and in the CPU consumption. Let stick on one of them, then: 11834.

As expected, the increase in CPU time because of the new modules run is quite limited:

```
The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.01%         0.00 ms/ev ->         1.45 ms/ev muonReducedTrackExtras
       added      +0.00%         0.00 ms/ev ->         0.09 ms/ev slimmedMuonTrackExtras
  ---------- ------------     --------                  ----       ------------
Job total:  20.6994 s/ev ==> 20.6467 s/ev
```

For the additional event size with the pt threshold currently set (4.5 GeV) the same results reported in the PR description apply (modulo some different compression factor due to the lower statistics used here):
- For AOD output:
```
Compare packed values
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->       287.1        287     NEWO   0.01     recoTrackExtras_muonReducedTrackExtras__RECO.
      0.0 ->       207.2        207     NEWO   0.00     SiStripClusteredmNewDetSetVector_muonReducedTrackExtras__RECO.
      0.0 ->       139.7        140     NEWO   0.00     SiPixelClusteredmNewDetSetVector_muonReducedTrackExtras__RECO.
      0.0 ->       249.0        249     NEWO   0.01     TrackingRecHitsOwned_muonReducedTrackExtras__RECO.
      0.0 ->       264.9        265     NEWO   0.01     recoTrackExtrasedmAssociation_muonReducedTrackExtras__RECO.
-------------------------------------------------------------
  4613449 ->     4614583       1134             0.0     ALL BRANCHES
```
- For miniAOD output
```
Compare packed values
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->       179.5        180     NEWO   0.18     SiStripClusteredmNewDetSetVector_slimmedMuonTrackExtras__RECO.
      0.0 ->       317.8        318     NEWO   0.31     recoTrackExtras_slimmedMuonTrackExtras__RECO.
      0.0 ->       126.2        126     NEWO   0.12     SiPixelClusteredmNewDetSetVector_slimmedMuonTrackExtras__RECO.
      0.0 ->       372.5        373     NEWO   0.36     TrackingRecHitsOwned_slimmedMuonTrackExtras__RECO.
-------------------------------------------------------------
   102561 ->      103558        997             1.0     ALL BRANCHES
```

perrotta commented:
@bendavid Maybe it is hidden somewhere in this long thread, but could you  please confirm that the added products have been tested to be sufficient for running the analysis jobs (i.e. nothing was forgotten)? Did you run an analyzer on the output produced with this PR to verify it?

bendavid commented:
Yes, I've explicitly tested refitting the tracks from MINIAOD.

Doing this requires an extra small producer to run at analysis level to copy the embedded reco::Tracks from the pat::Muon back into a separate collection, since the standard TrackRefitter takes a collection of reco::Tracks as input.

I can also commit this to a test directory somewhere if you prefer.

perrotta commented:
> Yes, I've explicitly tested refitting the tracks from MINIAOD.
> 
> Doing this requires an extra small producer to run at analysis level to copy the embedded reco::Tracks from the pat::Muon back into a separate collection, since the standard TrackRefitter takes a collection of reco::Tracks as input.
> 
> I can also commit this to a test directory somewhere if you prefer.

Thank you Josh, Yes, please do.

Even better if you could also add a simple unit test that keeps everything checked for correctness, also in case of future updates

bendavid commented:
Any preferences whether this goes in PhysicsTools/PatAlgos or RecoTracker/TrackProducer ? (since the test will run the TrackRefitter)

perrotta commented:
> Any preferences whether this goes in PhysicsTools/PatAlgos or RecoTracker/TrackProducer ? (since the test will run the TrackRefitter)

Uhm, not really: no any preference for me

bendavid commented:
Is there an equivalent to TrackRefitter that works with StandAlone Muon tracks and/or global tracks btw?
(my current test workflow only uses the inner tracks)

bendavid commented:
Ok, so I think we agreed to loosen the cut for AOD, but not MINIAOD.

Lowering it from pt>4.5 to pt>3.0 would bring the size increase for AOD from 0.16% to 0.25%.  Doing some final checks now that this does what is expected and then will update the PR.

Further including in AOD the extra content for pfMuons below 3GeV would increase that to ~0.6% as already tested previously (but I'm not sure to what extent these are used for analysis in BPH, etc)

bendavid committed:
loosen cut for muon track extras to 3 GeV for AOD (but maintain 4.5GeV cut for MINIAOD)

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31217/18101

- This PR adds an extra 192KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #30544, #30898, #31129
  - File  PhysicsTools/PatAlgos/python/slimming/slimmedTrackExtras_cff.py  modified in PR(s): #30544
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #30544, #31129
  - File  RecoMuon/Configuration/python/RecoMuon_EventContent_cff.py  modified in PR(s): #30544
  - File  RecoMuon/Configuration/python/RecoMuon_cff.py  modified in PR(s): #30544
  - File  RecoMuon/MuonIdentification/plugins/Module.cc  modified in PR(s): #30544

bendavid commented:
Cuts and description updated.

I would propose to defer adding the tests to the release to another pull request, because getting the refitting running also for the muon tracks required some hacking around for the moment.
(See discussion in https://hypernews.cern.ch/HyperNews/CMS/get/muon/1665.html)

cmsbuild commented:
Pull request #31217 was updated. @perrotta, @civanch, @jordan-martins, @chayanit, @wajidalikhan, @mdhildreth, @cmsbuild, @silviodonato, @franzoni, @kpedro88, @slava77, @jpata, @qliphy, @santocch, @fabiocos, @davidlange6 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

perrotta commented:
> Cuts and description updated.
> 
> I would propose to defer adding the tests to the release to another pull request, because getting the refitting running also for the muon tracks required some hacking around for the moment.
> (See discussion in https://hypernews.cern.ch/HyperNews/CMS/get/muon/1665.html)

Fine with me.
But please do it, eventually

cmsbuild commented:
+1
Tested at: 1caf73f136ce3c8716f7de164d266c247e566bcc
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/9074/summary.html
CMSSW: CMSSW_11_2_X_2020-09-01-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c4fcc/9074/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2609667
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2609639
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

perrotta commented:
+1
- It adds the extra collection to the AOD and miniAOD outputs as described in the PR description and follow up reviews
- Discussed in several forums, the latest thresholds (3 GeV for AOD and 4.5 GeV for miniAOD) were agreed by Physics and PPD during the Sep 2 PPD coordination meeting: corresponding event size increases (measured with the 11834.0 wf as approx 1.7 kB/evt for AOD and 1 kb/evt for miniAOD) are deemed acceptable
- Jenkins tests pass and show the additional collections in output, without any modification to the other reco outputs. The consistency of the added content was verified by the author: a unit test could be usefully added with a forthcoming PR to verify the consistency of those collections in case of possible future modifications

civanch commented:
+1

kpedro88 commented:
+upgrade

silviodonato commented:
+operations

silviodonato commented:
kind reminder @cms-sw/analysis-l2 @cms-sw/pdmv-l2

chayanit commented:
+1

silviodonato commented:
merge

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
